5.0 GA: update release notes, experimental features, roadmap (#5147)
pingcap/docs: experimental-features.md

---
title: TiDB Experimental Features
summary: Learn the experimental features of TiDB.
aliases: ['/tidb/dev/experimental-features-4.0/']
---

# TiDB Experimental Features

This document introduces the experimental features of TiDB in different versions. It is **NOT** recommended to use these features in the production environment.

## Stability

+ TiFlash limits the use of I/O resources by compressing or sorting data, alleviating the contention for I/O resources between background tasks and front-end data reading and writing (v5.0 experimental feature)

> + Improve the stability of the optimizer's choice of index (v5.0 experimental feature)

Reviewer: ```suggestion
+ Improve the stability of the optimizer's choice of an index (v5.0 experimental feature)
```

>     + Extend the statistics feature by collecting more information such as multi-column NDV (Number of Distinct Values), multi-column order dependency, and multi-column function dependency.

Author: ```suggestion
    + Extend the statistics feature by collecting the multi-column order dependency information.
```

    + Refactor the statistics module, including deleting the `TopN` value from `CMSKetch` and the histogram, and adding NDV information for histogram buckets of each table index.

> 

Author: ```suggestion
+ [hibernate-regions](/tikv-configuration-file.md#hibernate-regions-experimental). A Region that idles for a long time is automatically set as hibernated. This reduces the extra overhead caused by heartbeat messages between the Raft leader and the followers for idle Regions.

```

>     + Refactor the statistics module, including deleting the `TopN` value from `CMSKetch`, refactoring the `TopN` search logic and deleting the `TopN` information from the histogram, and create a histogram index to help maintain bucket NDV.

Reviewer: ```suggestion
    + Refactor the statistics module, including deleting the `TopN` value from `CMSKetch`, refactoring the `TopN` search logic, deleting the `TopN` information from the histogram, and create a histogram index to help maintain bucket NDV.
```

> 

Reviewer 2: This seems to missing:
* https://docs.pingcap.com/tidb/v5.0/system-variables#tidb_enable_async_commit-new-in-v500-rc

Author: This document has been updated. Please see the latest version in the preview environment I gave you earlier.

Reviewer 2: I meant it is missing from the experimental list here.

Author: I meant in the latest version, `tidb_enable_async_commit` is not experimental.

Reviewer 2: Ahh okay! Thanks.

Author: No problem.

+ Cascading Placement Rules feature. It is a replica rule system that guides PD to generate corresponding schedules for different types of data. By combining different scheduling rules, you can finely control the attributes of any continuous data range, such as the number of replicas, the storage location, the host type, whether to participate in Raft election, and whether to act as the Raft leader. See [Cascading Placement Rules](/configure-placement-rules.md) for details. (Introduced in v4.0)
+ Elastic scheduling feature. It enables the TiDB cluster to dynamically scale out and in on Kubernetes based on real-time workloads, which effectively reduces the stress during your application's peak hours and saves overheads. See [Enable TidbCluster Auto-scaling](https://docs.pingcap.com/tidb-in-kubernetes/stable/enable-tidb-cluster-auto-scaling) for details. (Introduced in v4.0)

> + [hibernate-regions](/tikv-configuration-file.md#hibernate-regions-experimental). A Region that idles for a long time is automatically set as hibernated. This reduces the extra overhead caused by heartbeat messages between the Raft leader and the followers for idle Regions.

Author: ```suggestion
```

## SQL

> 

Reviewer 2: User variables are also experimental: https://docs.pingcap.com/tidb/stable/user-defined-variables#user-defined-variables

They shouldn't be (because it's hard to avoid them), but I've checked with QA and this is currently correct.

Author: See [Line 30](https://github.com/pingcap/docs/pull/5147/files#diff-8c61c664e7ba1c50d254723c68d970c5bda812c90f8fbf544118057da2b4bdb0R30)

> + [hibernate-regions](/tikv-configuration-file.md#hibernate-regions-experimental). A Region idle for a long time is automatically set as hibernated. This reduces the extra overhead caused by heartbeat messages between the Raft leader and the followers for idle Regions.

Reviewer: ```suggestion
+ [hibernate-regions](/tikv-configuration-file.md#hibernate-regions-experimental). A Region that idles for a long time is automatically set as hibernated. This reduces the extra overhead caused by heartbeat messages between the Raft leader and the followers for idle Regions.
```

> + List Partition (v5.0 experimental feature)

Author: It's better. Updated.

+ List COLUMNS Partition (v5.0 experimental feature)
+ The expression index feature. The expression index is also called the function-based index. When you create an index, the index fields do not have to be a specific column but can be an expression calculated from one or more columns. This feature is useful for quickly accessing the calculation-based tables. See [Expression index](/sql-statements/sql-statement-create-index.md) for details. (v4.0 experimental feature)
+ [Generated Columns](/generated-columns.md).
+ [User-Defined Variables](/user-defined-variables.md).
+ [JSON data type](/data-type-json.md) and [JSON functions](/functions-and-operators/json-functions.md).
+ [View](/information-schema/information-schema-views.md).

## Configuration management

> + Persist the configuration parameters in PD, and support dynamically modifying the feature of configuration items. (v4.0 experimental feature)

Reviewer: "Persist" is intransitive when it means "exist" or "happen". 
```suggestion
+ Store configuration parameters persistently in PD, and support the feature of dynamically modifying configuration items. (v4.0 experimental feature)
```

+ [hibernate-regions](/tikv-configuration-file.md#hibernate-regions-experimental). A Region idle for a long time is automatically set as hibernated. This reduces the extra overhead caused by heartbeat messages between the Raft leader and the followers for idle Regions.

## SQL

+ List Partition (v5.0 experimental feature)
+ List COLUMNS Partition (v5.0 experimental feature)
+ The expression index feature. The expression index is also called the function-based index. When you create an index, the index fields do not have to be a specific column but can be an expression calculated from one or more columns. This feature is useful for quickly accessing the calculation-based tables. See [Expression index](/sql-statements/sql-statement-create-index.md) for details. (v4.0 experimental feature)
+ [Generated Columns](/generated-columns.md).
+ [User-Defined Variables](/user-defined-variables.md).
+ [JSON data type](/data-type-json.md) and [JSON functions](/functions-and-operators/json-functions.md).

> + [Global Kill](/sql-statements/sql-statement-kill.md).

Author: ```suggestion
```
